packages/asset-swapper/contracts/src/CoFiXSampler.sol

```diff
@@ -0,0 +1,226 @@
+import "./interfaces/ICoFiXController.sol";
+import "./interfaces/ICoFiXFactory.sol";
+import "./interfaces/ICoFiXPair.sol";
+import "./interfaces/INestOracle.sol";
+
+
+contract CoFiXSampler is
+    DeploymentConstants,
+    SamplerUtils
+{
+    /// @dev Gas limit for CoFiX calls.
+    uint256 constant private COFIX_CALL_GAS = 450e3; // 450k
+
+    struct OraclePrice {
+        uint256 ethAmount;
+        uint256 erc20Amount;
+        uint256 blockNum;
+        uint256 K;
+        uint256 theta;
+    }
+
+    /// @dev Sample sell quotes from CoFiX.
+    /// @param takerToken Address of the taker token (what to sell).
+    /// @param makerToken Address of the maker token (what to buy).
+    /// @param takerTokenAmounts Taker token sell amount for each sample.
+    /// @return makerTokenAmounts Maker amounts bought at each taker token
+    ///         amount.
+    function sampleSellsFromCoFiX(
+        address takerToken,
+        address makerToken,
+        uint256[] memory takerTokenAmounts
+    )
+        public
+        view
+        returns (uint256[] memory makerTokenAmounts)
+    {
+        // make sure it's not WETH-WETH
+        _assertValidPair(makerToken, takerToken);
+        uint256 numSamples = takerTokenAmounts.length;
+        makerTokenAmounts = new uint256[](numSamples);
+        ICoFiXPair takerTokenPair = takerToken == _getWethAddress() ?
+            ICoFiXPair(0) : _getCoFiXPair(takerToken);
+        ICoFiXPair makerTokenPair = makerToken == _getWethAddress() ?
+            ICoFiXPair(0) : _getCoFiXPair(makerToken);
+        for (uint256 i = 0; i < numSamples; i++) {
+            bool didSucceed = true;
+            if (makerToken == _getWethAddress()) {
+                // if converting to WETH, sell all of the token in the taker token pair
+                (didSucceed, bytes memory resultData) =
+                    address(takerTokenPair).staticcall.gas(COFIX_CALL_GAS)(
+                        abi.encodeWithSelector(
+                            ICoFiXPair(0).calcOutToken0.selector,
+                            takerTokenAmounts[i],
+                            _getOraclePrice(takerToken)
+                    ));
+                makerTokenAmounts[i] = abi.decode(resultData, (uint256[]))[0];
+            } else if (takerToken == _getWethAddress()) {
+                // if starting with WETH, sell all of the WETH in the maker token pair
+                (didSucceed, bytes memory resultData) =
+                    address(makerTokenPair).staticcall.gas(COFIX_CALL_GAS)(
+                        abi.encodeWithSelector(
+                            ICoFiXPair(0).calcOutToken1.selector,
+                            takerTokenAmounts[i],
+                            _getOraclePrice(makerToken)
+                    ));
+                makerTokenAmounts[i] = abi.decode(resultData, (uint256[]))[0];
+            } else {
+                // if it's converting a non-WETH token to another non-WETH token
```

Comment:
I think it's unlikely we'd want this case. From what I can see everything is ETH/X.

MultiHop should take care of this when it is X/Y rather then us having to do it here and in the bridge.

We do this for Kyber as all of Kyber trades are actually X->ETH->Y.